keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -231,15 +239,23 @@ type EndpointPort struct {
 
 ### EndpointMeta (Per EndpointSlice)
 EndpointMeta contains metadata applying to all endpoints contained in the EndpointSlice.
-- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.   
+- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.
 
 ### Topology (Per Endpoint)
 A new topology field (string to string map) is added to each endpoint. It can contain arbitrary topology information associated with the endpoint. If the EndpointSlice instance is derived from K8s service, the topology may contain following well known key:
 - **kubernetes.io/hostname**: the value indicates the hostname of the node where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/zone**: the value indicates the zone where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/region**: the value indicates the region where the endpoint is located. This should match the corresponding node label.
 
-If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology. 
+If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology.
+
+### Termination Grace Period (Per Endpoint)
+
+An optional field `TerminationGracePeriodSeconds` can be added to each Endpoint indicating the longest duration an endpoint is given to gracefully shutdown. This is analogous to `TerminationGracePeriodSeconds` in `PodSpec`. In most cases, the endpointslice controller will set `TerminationGracePeriodSeconds` on an `Endpoint` based on the termination grace period on `PodSpec`. External implementations of the `EndpointSlice` API can set this field arbitrarily based on specific implementation requirements.
+
+This field allows consumers of the API to know the termination grace period of endpoints (pods in this case) without having to watch Pods. A specific use-case for this field is the [graceful termination handler in the IPVS proxy](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/graceful_termination.go). Because deleting an IPVS backend (a.k.a IPVS real server) immediately deletes all its correpsonding conntrack entries, deleting a real server is disruptive during a pod's graceful termination period. Today, the IPVS proxy will only delete a real server if it sees that it has no connections or it has exceeded a hardcoded duration limit (15m). The `TerminationGracePeriodSeconds` field would remove the need for this hardcoded limit, improving overall performance of the proxy as it no longer needs to track stale backends.
```

Comment:
You have two options:
- set the backend weight to 0 which prevents new connections but allows existing ones to graceful terminate (this what we do today on endpoint deletion)
- delete the backend, which removes it and terminates the connection by clearing the conntrack and sending a RST on the next packet (which we only do today when there are no active connections to the backend)

Reply:
> The pod may choose to exit immediately or the node may go down, causing reduced grace. In that case, you will still be routing traffic to it as a realserver, when you could have stopped.

As @lbernail mentioned, for the IPVS proxier we reduce the weight of the backend to 0 so no new connections are created, so the grace period only applies to existing connections. The IPVS proxier also tracks connections by state, so we would remove the terminating endpoint early if we saw that all existing connections closed cleanly. In the few edge cases where the connection isn't closed properly, the real server would stay around. So at worse we would be leaving an existing connection open longer than it should up to `terminationGracePeriodSeconds`, but this is still better than the 15m duration today.

Reply:
I see, yeah that makes more sense. :)  Thanks for explaining.

You still have the kube-proxy restart problem, right?  Specifically

* Pod X is deleted
* 10 minute grace period starts
* X is removed from endpoints
* kube-proxy sets IPVS weight to 0
* kube-proxy restarts
* WATCH returns endpoints without X
* kube-proxy observes realserver X with weight 0 but has no idea if that is seconds old or days old

So it seems to lose information, and at some point needs to make a dubious assumption about how long to wait.

Reply:
yeah that's true -- for what it's worth the restart problem exists today since graceful termination handling is all done in local memory.

The solution I proposed https://github.com/kubernetes/enhancements/pull/1607#issuecomment-597632219 would fix this, but requires changing the behavior so that terminating pods are kept in the slice.